File: frontend/lib/api/overview.ts
import { apiGet } from "./client";

export interface NamedReturn {
  name: string;
  return_pct: number | null;
}

export interface AssetDistRow {
  [key: string]: unknown;
  broad_category: string;
  count: number;
  avg_1d: number | null;
  avg_7d: number | null;
  avg_30d: number | null;
  avg_90d: number | null;
  avg_1y: number | null;
  med_30d: number | null;
  med_90d: number | null;
}

export interface TopAmcRow {
  [key: string]: unknown;
  fund_house: string;
  schemes_count: number;
  avg_30d: number | null;
  avg_90d: number | null;
  avg_1y: number | null;
}

export interface OverviewStats {
  total_schemes: number;
  total_amcs: number;
  total_nav_records: number;
  min_date: string | null;
  max_date: string | null;
  asset_dist: AssetDistRow[];
  top_amcs: TopAmcRow[];
  best_cat: NamedReturn | null;
}

export interface OverviewKpis {
  total_schemes: number;
  latest_date: string | null;
  top_performer: NamedReturn | null;
  lag_performer: NamedReturn | null;
  advancers: number;
  decliners: number;
  unchanged: number;
  median_return: number;
  avg_return: number;
  best_cat: NamedReturn | null;
}

export interface MacroTrendPoint {
  nav_date: string;
  "Asset Class": string;
  "Indexed Performance": number;
}

export interface CategoryMatrixRow {
  [key: string]: unknown;
  "Asset Class": string;
  Category: string;
  Schemes: number;
  "Avg TER %": number | null;
  "Avg 1D %": number | null;
  "Avg 7D %": number | null;
  "Avg 30D %": number | null;
  "Avg 90D %": number | null;
  "Avg 1Y %": number | null;
  "52W High Gap %": number | null;
  "Top Fund (30D) %": number | null;
  "Bottom Fund (30D) %": number | null;
}

export const getOverviewStats = () => apiGet<OverviewStats>("/api/overview/stats");

export const getOverviewKpis = (planType: string, optionType: string = "All Options", startDate?: string, endDate?: string) =>
  apiGet<OverviewKpis>("/api/overview/kpis", {
    plan_type: planType,
    option_type: optionType,
    start_date: startDate,
    end_date: endDate,
  });

export const getMacroTrend = (startDate: string, endDate: string, planType: string, optionType: string = "All Options") =>
  apiGet<MacroTrendPoint[]>("/api/overview/macro-trend", {
    start_date: startDate,
    end_date: endDate,
    plan_type: planType,
    option_type: optionType,
  });

export const getCategoryMatrix = (broadCategory: string) =>
  apiGet<CategoryMatrixRow[]>("/api/overview/category-matrix", { broad_category: broadCategory });

File: backend/app/schemas/overview.py
import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


class NamedReturn(BaseModel):
    name: str
    return_pct: Optional[float] = None


class OverviewStats(BaseModel):
    """Backs GET /api/overview/stats -- db.get_market_overview_stats()."""

    total_schemes: int
    total_amcs: int
    total_nav_records: int
    min_date: Optional[datetime.date]
    max_date: Optional[datetime.date]
    asset_dist: List[Dict[str, Any]]  # broad_category, count, avg_1d/7d/30d/90d/1y, med_30d/90d
    top_amcs: List[Dict[str, Any]]  # fund_house, schemes_count, avg_30d/90d/1y
    best_cat: Optional[NamedReturn]


class OverviewKpis(BaseModel):
    """Backs GET /api/overview/kpis -- db.get_kpis()."""

    total_schemes: int
    latest_date: Optional[datetime.date]
    top_performer: Optional[NamedReturn]
    lag_performer: Optional[NamedReturn]
    advancers: int
    decliners: int
    unchanged: int
    median_return: float
    avg_return: float
    best_cat: Optional[NamedReturn]
